uaclient/security.py

```diff
@@ -0,0 +1,397 @@
+        return self.response["release_codename"]
+
+    @property
+    def status(self):
+        return self.response["status"]
+
+
+class CVE:
+    """Class representing CVE response from the SecurityClient"""
+
+    def __init__(self, client: UASecurityClient, response: "Dict[str, Any]"):
+        self.response = response
+        self.client = client
+
+    @property
+    def id(self):
+        return self.response.get("id", "UNKNOWN_CVE_ID").upper()
+
+    def _get_related_notices(self):
+        if hasattr(self, "notices"):
+            return self.notices
+        self.notices = {}
+        for notice_id in self.response.get("notices", []):
+            self.notices[notice_id] = self.client.get_notice(
+                notice_id=notice_id
+            )
+        return self.notices
+
+    @property
+    def notice_ids(self):
+        notices_md = self._get_related_notices()
+        return list(notices_md.keys())
+
+    @property
+    def title(self):
+        descr = self.response.get("description", "UNKNOWN_CVE_DESCRIPTION")
+        notices = self._get_related_notices()
+        if not notices:
+            return descr
+        # TODO(If multiple notices, which title to use?)
+        for key in sorted(notices):
+            return notices[key].get("title")
+        return descr
+
+    @property
+    def packages_status(self) -> "Dict[str, CVEPackageStatus]":
+        """Dict of package status dicts for the current Ubuntu series.
+
+        Top-level keys are source packages names and each value is a
+        CVEPackageStatus object
+        """
+        if hasattr(self, "_packages_status"):
+            return self._packages_status  # type: ignore
+        self._packages_status = {}
+        series = util.get_platform_info()["series"]
+        for package in self.response["packages"]:
+            for pkg_status in package["statuses"]:
+                if pkg_status["release_codename"] == series:
+                    self._packages_status[package["name"]] = CVEPackageStatus(
+                        pkg_status
+                    )
+        return self._packages_status
+
+
+class USN:
+    """Class representing USN response from the SecurityClient"""
+
+    def __init__(self, client: UASecurityClient, response: "Dict[str, str]"):
+        self.response = response
+        self.client = client
+
+    @property
+    def id(self):
+        return self.response.get("id", "UNKNOWN_USN_ID").upper()
+
+    @property
+    def title(self):
+        return self.response.get("title")
+
+
+def query_installed_pkg_sources() -> "Dict[str, Dict[str, Any]]":
+    """Return a dict of all source packages installed on the system.
+
+    The dict keys will be source package name: "krb5". The value of each
+    item will contain two keys:
+    - package-list: List of binary package names intalled related to the
+         source package
+    - version: the version string of the package installed
+    """
+    out, _err = util.subp(
```

Comment:
This command will return duplicate source packages here because each binary package derived from that source is listed. So for the call above we would have seen the following 6 lines for all the krb-derived binary pkgs:
```bash
krb5-locales,krb5,1.17-6ubuntu4.1
libgssapi-krb5-2,krb5,1.17-6ubuntu4.1
libk5crypto3,krb5,1.17-6ubuntu4.1
libkrb5-26-heimdal,heimdal,7.7.0+dfsg-1ubuntu1
libkrb5-3,krb5,1.17-6ubuntu4.1
libkrb5support0,krb5,1.17-6ubuntu4.1
```


All said though, we don't need to track the individual package-lists for anything at the moment because USNs and CVEs work in terms of the source package name, not binaries, so I'm modifying this query to only look at source pkg names.
I also added a check to ignore packages which are not installed. The previous dpkg-query didn't track db:Status-Status.

we now run
```bash
dpkg-query -f='${Source},${Version},${db:Status-Status}\n' -W
````

and ignore any line without "installed" status